File: data/fetch_urls_sequentially.rb
#!/usr/bin/env ruby

require_relative "urlsafe_name.rb"
require 'pty'

File.foreach("target_urls") do |url|
  wpr_output = "wpr/#{filesafe_name(url)}.wpr"
  cmd = "sudo ../analysis/web-page-replay/replay.py -i ../analysis/web-page-replay/deterministic.js --record #{wpr_output}"
  begin
    # Boot WPR
    puts "Starting WPR"
    (stdin, stdout, pid) = PTY.spawn( cmd )
    begin
      # Wait for it to complete its boot process:
      puts "Waiting for WPR to boot..."
      stdin.each do |line|
        puts line
        break if line =~ /HTTPS server started on/
      end
    rescue Errno::EIO
      puts "Errno:EIO error, but this probably just means " +
            "that the process has finished giving output"
    end

    puts "Fetching page #{url}"
    har_output = "har/#{filesafe_name(url)}.har"
    system("phantomjs --disk-cache=false netsniff.js #{url} > #{har_output}")

    puts "Shutting down WPR"
    # TODO(cs): figure out how to kill a root process using ruby's
    # Process.kill
    system("sudo kill -15 #{pid}")
  rescue PTY::ChildExited
    puts "The child process exited!"
  end
end

File: data/fetch_urls_sequentially.py
#!/usr/bin/env python

import os
import subprocess
import signal
import sys

import telemetry_web_page_replay as wpr
from telemetry_util import FileSafeName, exponential_backoff

phantomjs_path = "phantomjs"
if os.path.exists("/home/cs/local/bin/phantomjs"):
  phantomjs_path = "/home/cs/local/bin/phantomjs"


def fetch(url):
  print "Fetching ", url
  filename = FileSafeName(url)
  wpr_output = "wpr/" + filename + ".wpr"
  phantomjs_err = "wpr/" + filename + ".err"
  har_output = "har/" + filename + ".har"
  with wpr.ReplayServer(wpr_output, replay_options=["--record"]):
    subprocess.Popen(
      "%s --disk-cache=false --ignore-ssl-errors=true "
      "netsniff.js '%s'>'%s' 2>'%s'" %
      (phantomjs_path, url, har_output, phantomjs_err),
      shell=True)


if __name__ == '__main__':
  if os.geteuid() != 0:
    print "Must run as root"
    sys.exit(1)

  with open("target_urls") as f:
    for url in f:
      url = url.strip()
      exponential_backoff(fetch, url)
